docs/design/node-resource-reservation-design.md

```diff
@@ -0,0 +1,52 @@
+![annotation](images/node-resource-reservation-annotation.png) 
+### recognize pods which need to be reserved
+There are two ways to recognize these pods:
+* set annotation volcano.sh/reserveable: true in podgroup 
+which means all pods under podgroup are reserve tasks
+* set annotation volcano.sh/reserveable: true in pod
+which means this pod is reserve task
+### recognize pod max running time
+* set annotation volcano.sh/maximum-runtime: 500s in podgroup
+which means the podgroup will run for a maximum of 500 seconds
+* set annotation volcano.sh/maximum-runtime: 500s in pod
+which means this pod will run for a maximum of 500 seconds
+### reserve plugin
+#### configuration
+```
+- plugins:
+   - name: reserve
+     arguments:
+       reservelabels:
+       - nodeSelector:
+           business_type: ebook
+         startHour: 3
+         endHour: 4
+         resources:
+           cpu: "32"
+           memory: 64Gi
+         startReserveAgo: 2h
+         podNumToReserve: 10
+         cron: daily
+```
+In the configuration, reservelabels is consisted by nodeSelector which represent a node list, resources represent a list of resource reservation configuration.The overall meaning is that from 3 to 4 o'clock every day, 32 cpu, 64Gi memory need to be reserved and should start reserve 2h ago, if 10 reserve tasks are scheduled during reserve time range, stop reserve. which can save more resources for non-reserved tasks after 10 reserved tasks are scheduled during reserve period.
+#### OpenSession
+* make cache of ReserveNodesMap which is used to cache forbidden nodes to forbid non-reserved tasks to be scheduled on reserved nodes, the calculation algorithm is as follows: firstly, order the nodes desc by node idle. Node idle is consisted of node resource unused and the resource will be released in the future before reserve start time which is taken by the annotation of pod max running time. secondly, traverse the ordered nodes, accumulate the node allocatable resource, if the accumulated resource is less than the resource to be reserved, add the node to ReserveNodesMap which means the system will have the trend to reserve big resource other than many small resources.
+* make cache of reservedTaskPendingResource which is used to cache the accumulated resource of pending tasks
+* make cache of reservedTaskAllocatedResource which is used to cache the accumulated resource of allocated tasks
+* make cache of resourceIdle which is used to accumulate the node futureidle resource.
+* register plugin function PredicateFn
+* register event handler AllocateFunc and DeallocateFunc to dynamically update the cache of reservedTaskPendingResource, reservedTaskAllocatedResource and resourceIdle
+#### PredicateFn
+Predicate is used to restrict other pods to be scheduled on reserved nodes. Reserved nodes are filtered out from the list of nodes and will change dynamically. 
+* check if the task is a reserve task, if yes, permit the task to be scheduled on this node.
```

Comment:
Do you mean that if a node is selected by the label of the reserve plugin, then this node cannot schedule non-reserve tasks during the reserve time period? What if there is no reserve task in the reserve period, non-reserve tasks should wait until reserve time period ends?

Reply:
Not exactly. If the node idle resource(from resourceIdle cache) is larger than the reserve requirements(max(reservedTaskAllocatedResource + reservedTaskPendingResource, reserveResourcesConfig)), if yes, permit the non-reserved task to be scheduled on this node.